Number_theory.hpp

```diff
@@ -0,0 +1,261 @@
+#pragma once
+#include <iostream>
+#include <type_traits>
+#include <concepts>
+#include <vector>
+#include <cmath>
+namespace number_theory
+{
+    // Function to calculate the greatest common divisor (GCD) of two integers.
+    auto gcd(std::integral auto a, std::integral auto b) {
+        while (b != 0) {
+            int t = b;
+            b = a % b;
+            a = t;
+        }
+        return a;
+    }
+
+    // Function to check if a vector of numbers are pairwise coprime.
+    bool are_coprime(std::vector<std::size_t> numbers)
+    {
+        for (std::size_t i = 0; i < numbers.size(); i++)
+        {
+            for (std::size_t j = i + 1; j < numbers.size(); j++)
+            {
+                if (gcd(numbers[i], numbers[j]) != 1)
+                {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    /*finds num ^ exponent modulo mod, ensure that it is of O(log(exponent))
+     * complexity*/
+    template <std::size_t mod>
+    inline auto modular_pow(std::integral auto num, std::size_t exponent)
+    {
+        if (mod == 1) return 0;
+        decltype(num) result = 1;
+        num %= mod;
+
+        while (exponent > 0)
+        {
+            if (exponent % 2 == 1)
+                result = (result * num) % mod;
+            num = (num * num) % mod;
+            exponent /= 2;
+        }
+        return result;
+    }
+
+    /*finds (if exists) modular inverse of num,that is a number inv such that
+     * num * inv = 1 modulo mod*/
+    inline auto modular_inverse(std::integral auto num, std::size_t mod)
+    {
+        auto m = mod;
+
+        if (mod == 0)
+        {
+            throw std::invalid_argument("The mod is not an integer.");
+        }
+
+        if (gcd(num, mod) != 1)
+        {
+            throw std::invalid_argument("The numbers must be coprime.");
+        }
+
+
+        auto a = num % m;
+        auto m0 = m;
+        auto t = 0, q = 0;
+        auto x0 = 0, x1 = 1;
+
+        if (m == 1)
+            return 1;
+
+        while (a > 1) {
+            q = a / m;
+            t = m;
+            m = a % m;
+            a = t;
+            t = x0;
+            x0 = x1 - q * x0;
+            x1 = t;
+            if (m == 0)
+                break;
+        }
+
+        if (x1 < 0)
+            x1 += m0;
+
+        return x1;
+    }
+
+    /*find a solution x of ax = b modulo mod*/
+    template <std::size_t mod>
+    inline auto linear_congruence_solver(std::integral auto a, std::integral auto b)
+    {
+        if (mod == 0)
+        {
+            throw std::invalid_argument("mod cannot be zero.");
+        }
+
+        decltype(a) x;
+
+        if (b % a == 0 && mod % a == 0)
+        {
+            x = (b / a) % (mod / a);
+            return x;
+        }
+
+        if (gcd(a, mod) != 1)
+        {
+            throw std::invalid_argument("a and mod must be coprime.");
+        }

```

Comment:
this together with mod ==0 should be wrapped into sth like validate_args_for_congruence()